src/risedevtool/src/compose.rs

```diff
@@ -0,0 +1,127 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Generate docker compose yaml files for risedev components.
+
+use std::collections::HashMap;
+use std::process::Command;
+
+use anyhow::{anyhow, Result};
+use serde::Serialize;
+
+use crate::{
+    ComputeNodeConfig, ComputeNodeService, FrontendConfig, FrontendServiceV2, MetaNodeConfig,
+    MetaNodeService,
+};
+
+#[derive(Debug, Clone, Serialize, Default)]
+pub struct ComposeService {
+    pub image: String,
+    pub command: Vec<String>,
+    pub expose: Vec<String>,
+    pub ports: Vec<String>,
+    pub depends_on: Vec<String>,
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct ComposeFile {
+    pub version: String,
+    pub services: HashMap<String, ComposeService>,
+}
+
+/// Generate compose yaml files for a component.
+pub trait Compose {
+    fn compose(&self) -> Result<ComposeService>;
+}
+
+fn get_cmd_args(cmd: &Command) -> Result<Vec<String>> {
+    let mut result = vec![];
+    result.push(
+        cmd.get_program()
+            .to_str()
+            .ok_or_else(|| anyhow!("Cannot convert to UTF-8 string"))?
+            .to_string(),
+    );
+    for arg in cmd.get_args() {
+        result.push(
+            arg.to_str()
+                .ok_or_else(|| anyhow!("Cannot convert to UTF-8 string"))?
+                .to_string(),
+        );
+    }
+    Ok(result)
+}
+
+impl Compose for ComputeNodeConfig {
+    fn compose(&self) -> Result<ComposeService> {
+        let mut command = Command::new("/risingwave/bin/compute-node");
+        ComputeNodeService::apply_command_args(&mut command, self)?;
+        let command = get_cmd_args(&command)?;
+
+        Ok(ComposeService {
+            image: "ghcr.io/singularity-data/risingwave:latest".into(),
+            command,
+            expose: vec![self.port.to_string(), self.exporter_port.to_string()],
+            depends_on: self
+                .provide_meta_node
+                .as_ref()
+                .unwrap()
+                .iter()
+                .map(|x| x.id.clone())
+                .collect(),
+            ..Default::default()
+        })
+    }
+}
+
+impl Compose for MetaNodeConfig {
+    fn compose(&self) -> Result<ComposeService> {
+        let mut command = Command::new("/risingwave/bin/meta-node");
+        MetaNodeService::apply_command_args(&mut command, self)?;
+        let command = get_cmd_args(&command)?;
+
+        Ok(ComposeService {
+            image: "ghcr.io/singularity-data/risingwave:latest".into(),
+            command,
+            expose: vec![
+                self.port.to_string(),
+                self.exporter_port.to_string(),
+                self.dashboard_port.to_string(),
+            ],
+            ..Default::default()
+        })
+    }
+}
+
+impl Compose for FrontendConfig {
+    fn compose(&self) -> Result<ComposeService> {
+        let mut command = Command::new("/risingwave/bin/frontend-v2");
```

Comment:
```suggestion
        let mut command = Command::new(FRONTEND_BIN_PATH);
```

Reply:
In https://github.com/singularity-data/risingwave/pull/2299/, I've specified the path of binary in Dockerfile, so we won't need to specify here in the future.